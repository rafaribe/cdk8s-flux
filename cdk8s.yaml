imports:
  - k8s@1.16.0
