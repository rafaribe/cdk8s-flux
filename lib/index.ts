import { Construct } from 'constructs';
import {
  ClusterRole,
  ClusterRoleBinding,
  Deployment,
  Namespace,
  Probe,
  Quantity,
  Secret,
  Service,
  ServiceAccount,
  Volume,
} from '../imports/k8s';

export interface FluxOptions {
  /**
   * The image for flux (docker hub only)
   */
  readonly image: string;
  /**
   * The tag for the flux image
   */
  readonly tag: string;

  /**
   * The name for the flux container
   */
  readonly name: string;

  /**
   * The Namespace that should be used for the resources
   */
  readonly ns: string;

  /**
   * Number of replicas.
   *
   * @default 1
   */
  readonly replicas?: number;

  readonly memcachedTag?: string;

  /**
   * Flux Arguments
   *
   */
  readonly arguments: string[];
}

export class Flux extends Construct {
  constructor(scope: Construct, constructId: string, options: FluxOptions) {
    super(scope, constructId);
    const label = { name: options.name };

    const limits = {
      memory: Quantity.fromString('256Mi'),
      cpu: Quantity.fromString('150m'),
    };

    const fluxPort = 3030;
    const memcachedImageTag = options.memcachedTag ?? "1.5.2";

    new Namespace(this, options.ns, {
      metadata: {
        name: options.ns,
        labels: label,
      },
    });

    const serviceAccount = new ServiceAccount(this, options.name + 'sa', {
      metadata: {
        labels: label,
        name: options.name,
        namespace: options.ns,
      },
    });

    const clusterRole = new ClusterRole(this, options.name + '-cr', {
      metadata: {
        labels: label,
        name: options.name,
      },
      rules: [
        {
          apiGroups: ['*'],
          resources: ['*'],
          verbs: ['*'],
        },
        {
          nonResourceURLs: ['*'],
          verbs: ['*'],
        },
      ],
    });

    new ClusterRoleBinding(this, options.name + '-crb', {
      metadata: {
        labels: label,
        name: options.name,
      },
      roleRef: {
        apiGroup: 'rbac.authorization.k8s.io',
        kind: clusterRole.kind,
        name: clusterRole.name,
      },
      subjects: [
        {
          kind: serviceAccount.kind,
          name: serviceAccount.name,
          namespace: options.ns,
        },
      ],
    });

    const probe: Probe = {
      httpGet: {
        port: fluxPort,
        path: 'api/flux/v6/identity.pub',
      },
      initialDelaySeconds: 5,
      timeoutSeconds: 5,
    };
    const fluxGitDeploy = options.name + '-git-deploy';

    const volumes: Volume[] = [
      {
        name: 'git-key',
        secret: {
          secretName: fluxGitDeploy,
          defaultMode: 0o400,
        },
      },
      {
        name: 'git-keygen',
        emptyDir: { medium: 'Memory' },
      },
    ];

    new Secret(this, options.name + '-secret', {
      metadata: {
        name: fluxGitDeploy,
        namespace: options.ns,
      },
      type: 'Opaque',
    });

    new Deployment(this, options.name + '-dp', {
      metadata: {
        namespace: options.ns,
        name: options.name,
      },
      spec: {
        replicas: 1,
        selector: {
          matchLabels: label,
        },
        strategy: {
          type: 'Recreate',
        },
        template: {
          metadata: {
            labels: label,
          },
          spec: {
            serviceAccountName: serviceAccount.name,
            volumes: volumes,
            containers: [
              {
                name: options.name,
                image: options.image + ':' + options.tag,
                imagePullPolicy: 'IfNotPresent',
                ports: [{ containerPort: fluxPort }],
                resources: {
                  limits: limits,
                },
                livenessProbe: probe,
                readinessProbe: probe,
                volumeMounts: [
                  {
                    name: volumes[0].name,
                    mountPath: '/etc/fluxd/ssh',
                    readOnly: true,
                  },
                  {
                    name: volumes[1].name,
                    mountPath: '/etc/fluxd/keygen',
                  },
                ],
                args: options.arguments,
              },
            ],
          },
        },
      },
    });

    const memcachedName = 'memcached';
    const memcachedLabels = {
      name: memcachedName,
    };
    const memcachedPort: number = 11211;
    new Deployment(this, options.name + '-memcached', {
      metadata: {
        name: memcachedName,
        namespace: options.ns,
      },
      spec: {
        replicas: 1,
        selector: {
          matchLabels: memcachedLabels,
        },
        template: {
          metadata: {
            labels: memcachedLabels,
          },
          spec: {
            containers: [
              {
                name: memcachedName,
                image: 'memcached:' + memcachedImageTag,
                args: ['-m 512', '-I 5m', '-p ' + memcachedPort],
                ports: [
                  {
                    name: 'clients',
                    containerPort: memcachedPort,
                  },
                ],
                securityContext: {
                  runAsGroup: memcachedPort,
                  runAsUser: memcachedPort,
                  allowPrivilegeEscalation: false,
                },
              },
            ],
          },
        },
      },
    });
    new Service(this, memcachedName + '-svc', {
      metadata: {
        name: memcachedName,
        namespace: options.name,
      },
      spec: {
        ports: [
          {
            port: memcachedPort,
            name: memcachedName,
          },
        ],
        selector: {
          ['name']: memcachedName,
        },
      },
    });
  }
}
