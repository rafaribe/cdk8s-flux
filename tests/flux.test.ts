import { Testing, Chart } from 'cdk8s';
import { Flux } from '../lib';
import { expect, test } from '@jest/globals';
test('minimal configuration', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'test');

  // WHEN
  new Flux(chart, 'flux', {
    ns: 'flux',
    name: 'flux',
    image: 'fluxcd/flux',
    tag: '1.19.0',
    replicas: 1,
    arguments: [
      '--memcached-service=',
      '--ssh-keygen-dir=/etc/fluxd/keygen',
      '--git-url=git@github.com:rafaribe/cdk8s-k3s-gitops.git',
      '--git-branch=master',
      '--git-path=cluster',
      '--git-label=flux',
      '--git-user=flux',
      '--git-email=flux@rafaribe.com',
      '--git-poll-interval=5m',
      '--sync-garbage-collection',
    ],
  });
  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});
