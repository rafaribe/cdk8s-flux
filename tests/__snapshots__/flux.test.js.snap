// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal configuration 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Namespace",
    "metadata": Object {
      "labels": Object {
        "name": "flux",
      },
      "name": "flux",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "ServiceAccount",
    "metadata": Object {
      "labels": Object {
        "name": "flux",
      },
      "name": "flux",
      "namespace": "flux",
    },
  },
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "ClusterRole",
    "metadata": Object {
      "labels": Object {
        "name": "flux",
      },
      "name": "flux",
    },
    "rules": Array [
      Object {
        "apiGroups": Array [
          "*",
        ],
        "resources": Array [
          "*",
        ],
        "verbs": Array [
          "*",
        ],
      },
      Object {
        "nonResourceURLs": Array [
          "*",
        ],
        "verbs": Array [
          "*",
        ],
      },
    ],
  },
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "ClusterRoleBinding",
    "metadata": Object {
      "labels": Object {
        "name": "flux",
      },
      "name": "flux",
    },
    "roleRef": Object {
      "apiGroup": "rbac.authorization.k8s.io",
      "kind": "ClusterRole",
      "name": "flux",
    },
    "subjects": Array [
      Object {
        "kind": "ServiceAccount",
        "name": "flux",
        "namespace": "flux",
      },
    ],
  },
  Object {
    "apiVersion": "v1",
    "kind": "Secret",
    "metadata": Object {
      "name": "flux-git-deploy",
      "namespace": "flux",
    },
    "type": "Opaque",
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "flux",
      "namespace": "flux",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "flux",
        },
      },
      "strategy": Object {
        "type": "Recreate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "flux",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "args": Array [
                "--memcached-service=",
                "--ssh-keygen-dir=/etc/fluxd/keygen",
                "--git-url=git@github.com:rafaribe/cdk8s-k3s-gitops.git",
                "--git-branch=master",
                "--git-path=cluster",
                "--git-label=flux",
                "--git-user=flux",
                "--git-email=flux@rafaribe.com",
                "--git-poll-interval=5m",
                "--sync-garbage-collection",
              ],
              "image": "fluxcd/flux:1.19.0",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "httpGet": Object {
                  "path": "api/flux/v6/identity.pub",
                  "port": 3030,
                },
                "initialDelaySeconds": 5,
                "timeoutSeconds": 5,
              },
              "name": "flux",
              "ports": Array [
                Object {
                  "containerPort": 3030,
                },
              ],
              "readinessProbe": Object {
                "httpGet": Object {
                  "path": "api/flux/v6/identity.pub",
                  "port": 3030,
                },
                "initialDelaySeconds": 5,
                "timeoutSeconds": 5,
              },
              "resources": Object {
                "limits": Object {
                  "cpu": "150m",
                  "memory": "256Mi",
                },
              },
              "volumeMounts": Array [
                Object {
                  "mountPath": "/etc/fluxd/ssh",
                  "name": "git-key",
                  "readOnly": true,
                },
                Object {
                  "mountPath": "/etc/fluxd/keygen",
                  "name": "git-keygen",
                },
              ],
            },
          ],
          "serviceAccountName": "flux",
          "volumes": Array [
            Object {
              "name": "git-key",
              "secret": Object {
                "defaultMode": 256,
                "secretName": "flux-git-deploy",
              },
            },
            Object {
              "emptyDir": Object {
                "medium": "Memory",
              },
              "name": "git-keygen",
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "memcached",
      "namespace": "flux",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "memcached",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "memcached",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "args": Array [
                "-m 512",
                "-I 5m",
                "-p 11211",
              ],
              "image": "memcached:undefined",
              "name": "memcached",
              "ports": Array [
                Object {
                  "containerPort": 11211,
                  "name": "clients",
                },
              ],
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "runAsGroup": 11211,
                "runAsUser": 11211,
              },
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "memcached",
      "namespace": "flux",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "memcached",
          "port": 11211,
        },
      ],
      "selector": Object {
        "name": "memcached",
      },
    },
  },
]
`;
